ide/git/src/org/netbeans/modules/git/ui/history/DefaultGitHyperlinkProvider.java

```diff
@@ -0,0 +1,126 @@
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Map;
+import java.util.regex.Pattern;
+import java.util.stream.IntStream;
+import org.netbeans.libs.git.GitClient;
+import org.netbeans.libs.git.GitException;
+import org.netbeans.libs.git.GitRemoteConfig;
+import org.netbeans.libs.git.GitRepository;
+import org.netbeans.libs.git.GitURI;
+import org.netbeans.libs.git.progress.ProgressMonitor.DefaultProgressMonitor;
+import org.netbeans.modules.versioning.util.VCSHyperlinkProvider;
+import org.openide.awt.StatusDisplayer;
+import org.openide.util.Exceptions;
+import org.openide.util.NbBundle.Messages;
+import org.openide.util.RequestProcessor;
+import org.openide.util.lookup.ServiceProvider;
+
+/**
+ * Looks for '#123' patterns in text (e.g commit messages) and tries to link
+ * them to a fitting web page.
+ * 
+ * <p>The repository host is identified by inspecting the registered remotes
+ * of the local git repository.
+ * 
+ * @author mbien
+ */
+@Messages({
+     "# {0} - issue ID",
+     "TT_DefaultGitHyperlinkProvider.link=Click to open {0} in Browser",
+     "MSG_DefaultGitHyperlinkProvider.link.failed=Could not resolve issue link."
+ })
+@ServiceProvider(service=VCSHyperlinkProvider.class, position = Integer.MAX_VALUE)
+public class DefaultGitHyperlinkProvider extends VCSHyperlinkProvider {
+    
+    private static final Pattern ids = Pattern.compile("#[0-9]+");
+
+    @Override
+    public int[] getSpans(String text) {
+        return ids.matcher(text).results()
+                                .flatMapToInt(r -> IntStream.of(r.start(), r.end()))
+                                .toArray();
+    }
+
+    @Override
+    public String getTooltip(String text, int offsetStart, int offsetEnd) {
+        return Bundle.TT_DefaultGitHyperlinkProvider_link(text.substring(offsetStart, offsetEnd));
+    }
+
+    @Override
+    public void onClick(File file, String text, int offsetStart, int offsetEnd) {
+        Path path = file.toPath();
+        while (!Files.isDirectory(path) || Files.notExists(path.resolve(".git"))) {
+            path = path.getParent();
+            if (path == null) {
+                return;
+            }
+        }
+        Path repo = path;
+        RequestProcessor.getDefault().post(() -> {
+            openIssue(repo, text.substring(offsetStart + 1, offsetEnd));
+        });
+    }
+
+    private static void openIssue(Path repo, String id) {
+        try (GitClient client = GitRepository.getInstance(repo.toFile()).createClient()) {
+            Map<String, GitRemoteConfig> remotes = client.getRemotes(new DefaultProgressMonitor());
+            GitRemoteConfig origin = remotes.get("origin");
+            if (origin == null) {
+                origin = remotes.values().stream()
+                                         .filter(r -> r.getRemoteName().startsWith("origin/"))
+                                         .findFirst().orElse(null);
```

Comment:
Out of curiosity - is there really a pattern to name remotes "origin/whatever"? I knew that pattern from branch names, but not for remotes. I see no problem though.

Another thing I'm wondering is how to handle this case: On github I normally clone a repository I want to contribute to, then I create a local clone. At this point `origin` points to my repository. I then add a second remote `upstream`, that points to original repository. The issues I'm interested to are referred to by the url behind `upstream`, not the ones behind `origin`.

Reply:
Yes, almost all the projects I'm working on, including all NetBeans' ones, are set up locally like this for me too - `origin` as my fork, `upstream` as original repository (and, off-topic, with push disabled on upstream).

Reply:
I am actually not sure, I tested a few projects and the maven repo had for example:
```
$ git branch -r
  origin/HEAD -> origin/master
  origin/MNG-4463
  origin/MNG-4645
```
I had the same thought - what if someone uses origin differently. How about adding a well known origin name which is used just as a marker for NB?

e.g: `nb-origin`? If NB sees this it uses it instead of `origin`?

Reply:
(I have origin pointed at the root repo for pull, for push I use the full URL - I know this is not common but this is done to avoid mistakes, e.g I just pushed with `git push -f git@github.com:mbien/netbeans nb-upgrade-module-cleanup:nb-upgrade-module-cleanup` - everything is in that line and it will always work no matter how the repo is set up)

Reply:
@neilcsmith-net how common is `upstream`?

should I do this order:
`origin4nb`
`upstream`
`origin`
`origin/*`

Reply:
I've no idea, but given there are already two people who've looked at this say we do it ... :shrug: 

>  for push I use the full URL - I know this is not common but this is done to avoid mistakes,

Yes, absolutely!  Should be more common.  First thing I do after adding any upstream remote is `git remote set-url --push upstream DISABLED`.  Then use the full URL when absolutely required.